Improve loading

hatemhosny opened the pull request:
<!--
  For Work In Progress Pull Requests, please use the Draft PR feature,
  see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

  For a timely review/response, please avoid force-pushing additional
  commits if your PR already received reviews or comments.

  Before submitting a Pull Request, please ensure you've done the following:
  - 📖 Read the LiveCodes Contributing Guide: https://github.com/live-codes/livecodes/blob/HEAD/CONTRIBUTING.md.
  - 📖 Read the LiveCodes Code of Conduct: https://github.com/live-codes/livecodes/blob/HEAD/CODE_OF_CONDUCT.md.
  - 👷‍♀️ Create small PRs. In most cases, this will be possible.
  - ✅ Provide tests for your changes.
  - 📝 Use descriptive commit messages.
  - 📗 Update any related documentation and include any relevant screenshots.
  - 🌐 Use `window.deps.translateString` in .ts files and add `data-i18n` attributes in .html files to mark strings that needs to be translated.
-->

## What type of PR is this? (check all applicable)

- [ ] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [x] ♻️ Code Refactor
- [x] 🔥 Performance Improvements
- [x] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert
- [ ] 🌐 Internationalization / Translation

## Description

This PR improves loading new projects.

A project config can be generated from a variety of sources. Some of the sources are already available when loading (e.g. SDK config, params, compressed code in `x` param, etc), while others need to be imported (e.g. template, import URL, config content URL, etc)

These sources were mixed in [`importExternalContent`](https://github.com/live-codes/livecodes/blob/ed4c7f24cece43c6218184c581ce2e621e8262df/src/livecodes/core.ts#L5241). So, [`initializePlayground`](https://github.com/live-codes/livecodes/blob/ed4c7f24cece43c6218184c581ce2e621e8262df/src/livecodes/core.ts#L5415) started loading the UI then followed by a flash of changing content when `importExternalContent` runs.
In addition, a notification of "Loading Content" appeared even when the content was already available.

This PR now handles all already available sources in the initial load and keeps only remote content to `importExternalContent`.

@BassemHalim whould you please review this, which is very much related to your PR #819

hatemhosny commented:
This PR drops a very old (and undocumented) feature that allows the playground URL hash to act as an [import URL](https://livecodes.io/docs/features/import). This has been replaced by the `x` query param before the first available [permanent URL version](https://v3.livecodes.io/).

This kept conflicting with the new hash params introduced in #819 and was difficult (and not worth it) to differentiate with this refactor.

BassemHalim commented:
Hi Dr.@hatemhosny 
So I reviewed the PR and everything looks good. My only comment is if we need the new sdkConfig parameter in `importExternalContent`; the only time it is passed is in `initializePlayground` and in that function we do the following
```ts 
//core.ts:5410
 ...
  const sdkConfig = importCompressedCode(params.config ?? '');
  const initialConfig = { ...codeImportConfig, ...appConfig, ...sdkConfig };// <== includes sdkConfig
 ...
  setConfig(buildConfig({ ...getConfig(), ...userConfig, ...initialConfig })); // <== includes initialConfig
 ...
 importExternalContent({
    config: getConfig(), // <== this should have the fields in initialConfig so we shouldn't really need sdkConfig
    sdkConfig,
    configUrl: params.config,
    template: params.template,
    importUrl: Object.keys(codeImportConfig).length > 0 ? '' : params.x, // do not re-import compressed code
  }).then(async (contentImported) => {
```
so I think the props in `sdkConfig` should already be in `config` and we wouldn't need that new parameter.

hatemhosny commented:
Thank you @BassemHalim for the review.

> so I think the props in sdkConfig should already be in config and we wouldn't need that new parameter.

Yes, `sdkConfig` was already added. However, it is passed separately to `importExternalContent` to preserve precedence when [multiple sources](https://dev.livecodes.io/docs/sdk/js-ts#multiple-sources) are used. SDK config properties should override those in `template` and `import`. We need `sdkConfig` to be able to [do this](https://github.com/live-codes/livecodes/blob/9cc699279882da9b4e9f57ca06083bd263b83332/src/livecodes/core.ts#L5350-L5361):

```ts
  await loadConfig(
    buildConfig({
      ...config,
      ...templateConfig,
      ...importUrlConfig,
      ...configUrlConfig,
      ...sdkConfig,
      ...contentUrlConfig,
    }),
    parent.location.href,
    false,
  );
```

We cannot use `config` for the override because this also has the `defaultConfig` (from [`buildConfig`](https://github.com/live-codes/livecodes/blob/9cc699279882da9b4e9f57ca06083bd263b83332/src/livecodes/config/build-config.ts#L29)).

Example:

```ts
const options: EmbedOptions = {
  template: 'react',
  config: {
    activeEditor: 'markup',
    markup: {
      language: 'html',
      content: '<p>Hello World (from config)</p>',
    },
  },
  params: {
    console: 'open',
  },
};
```

demo: https://livecodes.io/?x=id/i8628qhyin6

> (P.S. I am still looking at the failing tests)

These were related to a feature to use hash as import url (now replaced by `x` param).
This was an old feature which was replaced by the query param very early on and did not even make it to the docs.
I removed this feature because it kept clashing with the new hash params.
However, there are tests that depend on this. So, I thought maybe someone else might have also been using it.
It felt bad to change the tests just to pass CI, so I reverted the change, and added a workaround to stay  backward-compatible!
Now the tests pass.

This PR still needs some tests for different scenarios and different sources.
I will add these before I merge it.

BassemHalim commented:
oh ok I see what you mean.  
I want to point out  one thing, so if a template was used and the config param overwrites it, it would only partially overwrite it for example here [https://livecodes.io/?x=id/t879rd4df3b](https://livecodes.io/?x=id/t879rd4df3b) if we set template to markdown then overwrite the markup with html it would overwrite the markup only and not the css or js. This still technically falls within the overwrite rules and is valid but I just wanted to point that out in case that is not the intended behavior.

hatemhosny commented:
Thank you @BassemHalim 

Yes, this is indeed the intended behaviour.
In your example, we may use the setup (e.g. styles, importmaps, scripts, ...etc) and just add content.
This can be significant (e.g. [Shadcn UI template](https://github.com/live-codes/livecodes/blob/ed4c7f24cece43c6218184c581ce2e621e8262df/src/livecodes/templates/starter/shadcn-ui-starter.ts#L149-L213))

Also this allows just adding query params to a share URL to modify some behaviours. e.g. `mode=result` or `console=open`

hatemhosny commented:
added [e2e tests](https://github.com/live-codes/livecodes/blob/77854e812284d47f63fcbe3ca2db058fabc91497/e2e/specs/sdk-options.spec.ts) for SDK options.

I think this is ready now to merge.
